spras/interval.py

```diff
@@ -0,0 +1,234 @@
+"""
+
+import tokenize
+from enum import Enum
+from io import BytesIO
+from typing import Any, ClassVar, Optional, Self, cast
+
+from pydantic import model_serializer, model_validator
+from pydantic.dataclasses import dataclass
+
+
+class Operand(Enum):
+    LT = "<"
+    LTE = "<="
+    EQ = "="
+    GTE = ">="
+    GT = ">"
+
+    @classmethod
+    def from_str(cls, string: str) -> Optional[Self]:
+        return next((enum for enum in list(cls) if enum.value == string), None)
+
+    def is_closed(self) -> bool:
+        """Whether this is a closed inequality. We consider = to be closed."""
+        match self:
+            case Operand.LTE: return True
+            case Operand.EQ: return True
+            case Operand.GTE: return True
+        return False
+
+    def as_closed(self):
+        """Closes an operand. Eq does not get modified."""
+        match self:
+            case Operand.LT: return Operand.LTE
+            case Operand.GT: return Operand.GTE
+        return self
+
+    def as_opened(self):
+        """Opens an operand. Eq does not get modified."""
+        match self:
+            case Operand.LTE: return Operand.LT
+            case Operand.GTE: return Operand.GT
+        return self
+
+    def with_closed(self, closed: bool): return self.as_closed() if closed else self.as_opened()
+
+    def compare(self, left, right) -> bool:
+        match self:
+            case Operand.LT: return left < right
+            case Operand.LTE: return left <= right
+            case Operand.EQ: return left == right
+            case Operand.GTE: return left >= right
+            case Operand.GT: return left > right
+
+    def flip(self):
+        match self:
+            case Operand.LT: return Operand.GT
+            case Operand.LTE: return Operand.GTE
+            case Operand.EQ: return Operand.EQ
+            case Operand.GTE: return Operand.LTE
+            case Operand.GT: return Operand.LT
+
+    @classmethod
+    def combine(cls, left: Self, right: Self):
+        """Combines two operands, returning None if the operands don't combine well."""
+        match (left, right):
+            case (Operand.LTE, Operand.LTE): return Operand.LTE
+            case (Operand.LT, Operand.LTE): return Operand.LT
+            case (Operand.LTE, Operand.LT): return Operand.LT
+            case (Operand.LT, Operand.LT): return Operand.LT
+            case (Operand.EQ, op): return op
+            case (op, Operand.EQ): return op
+            case (Operand.GTE, Operand.GTE): return Operand.GTE
+            case (Operand.GT, Operand.GTE): return Operand.GT
+            case (Operand.GTE, Operand.GT): return Operand.GT
+            case (Operand.GT, Operand.GT): return Operand.GT
+        return None
+
+@dataclass
+class Interval:
+    EMPTY_STRING: ClassVar[str] = "{empty interval}"
+
+    lower: Optional[float]
+    upper: Optional[float]
+    lower_closed: bool
+    upper_closed: bool
+
+    def mem(self, num: float) -> bool:
```

Comment:
What does `mem` stand for?

Reply:
If `num` is inside the interval: (we should swap this out with a library, and I'll check out the library from the below comment), but this should be `in`.